src/MarkdownWriter/ModulePageWriter.cs

```diff
@@ -0,0 +1,146 @@
+
+        public ModulePageWriter(MarkdownWriterSettings settings)
+        {
+            if (string.IsNullOrEmpty(settings.DestinationPath))
+            {
+                throw new ArgumentNullException("destinationpath");
+            }
+
+            _modulePagePath = settings.DestinationPath;
+            _encoding = settings.Encoding;
+            sb = Constants.StringBuilderPool.Get();
+        }
+
+        internal FileInfo Write(Collection<CommandHelp> helpItems)
+        {
+            if (helpItems.Count < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(helpItems));
+            }
+
+            // Help items in one module page have the same module name, locale and module GUID.
+            // So we can safely just get these values from the first object.
+            string moduleName = helpItems[0].ModuleName;
+            string localeString = helpItems[0].Locale.ToString();
+
+            string? helpModuleGuid = helpItems[0].ModuleGuid?.ToString();
+            string moduleGuid = helpModuleGuid is null ? Constants.FillInGuid : helpModuleGuid;
+
+            FileInfo modulePage = new FileInfo(_modulePagePath);
+
+            if (!string.Equals(modulePage.Extension, ".md", StringComparison.OrdinalIgnoreCase))
+            {
+                _modulePagePath = $"{_modulePagePath}\\{moduleName}.md";
+            }
+            else
+            {
+                if (!modulePage.Exists)
+                {
+                    DirectoryInfo? currentDir = modulePage.Directory;
+
+                    if (currentDir is not null)
+                    {
+                        bool currentDirExists = currentDir.Exists;
+
+                        while (!currentDirExists)
+                        {
+                            try
+                            {
+                                if (currentDir?.FullName is not null)
+                                {
+                                    Directory.CreateDirectory(currentDir.FullName);
+                                }
+                                break;
+                            }
+                            catch (DirectoryNotFoundException)
+                            {
+                                currentDir = currentDir?.Parent;
+                            }
+                        }
+                    }
+                }
+            }
+
+            using StreamWriter mdFileWriter = new(_modulePagePath, append: false, _encoding);
+
+            WriteHeader(moduleName, localeString, moduleGuid);
+
+            sb.AppendLine();
+
+            WriteModuleBlock(moduleName);
+
+            List<string> commandNames = new();
+
+            foreach (var help in helpItems)
+            {
+                commandNames.Add(help.Title);
+            }
+
+            WriteCmdletBlock(commandNames);
+
+            mdFileWriter.Write(sb.ToString());
+
+            Constants.StringBuilderPool.Return(sb);
```

Comment:
You may not return `sb` in case error happens before this line. Using `IDisposable` is better.